File: c/driver/postgresql/redshift_auth.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "redshift_auth.h"

#include <aws/core/Aws.h>
#include <aws/core/auth/AWSCredentialsProvider.h>
#include <aws/core/client/ClientConfiguration.h>
#include <aws/redshift/RedshiftClient.h>
#include <aws/redshift/model/GetClusterCredentialsRequest.h>

#include <memory>
#include <string>

#include "driver/common/utils.h"

namespace adbcpq {

AwsClientSingleton::AwsClientSingleton() {
    static bool initialized = false;
    if (!initialized) {
        Aws::InitAPI(options_);
        initialized = true;
    }
}

AwsClientSingleton::~AwsClientSingleton() {
  Aws::ShutdownAPI(options_);
}

class AwsAuthClient::Impl {
 public:
  Impl() {
    // Meyer's Singleton
    AwsClientSingleton::Instance();
  }

  ~Impl() {
    // no-op
  }

  Status GetRedshiftCredentials(const AwsAuthSettings& settings,
                                RedshiftCredentials& credentials,
                                struct AdbcError* error) const {
    try {
      Aws::Auth::AWSCredentials aws_credentials;
      if (settings.profile.has_value()) {
        auto status =
            GetCredentialsFromProfile(settings.profile.value(), aws_credentials, error);
        if (!status.ok()) return status;
      } else {
        if (!settings.access_key_id.has_value() ||
            !settings.secret_access_key.has_value()) {
          SetError(error, "%s",
                   "[aws] Both access_key_id and secret_access_key must be provided when "
                   "not using AWS profile");
          return Status::InvalidArgument(
              "Both access_key_id and secret_access_key must be provided when not using "
              "AWS profile");
        }
        aws_credentials = Aws::Auth::AWSCredentials(settings.access_key_id.value(),
                                                    settings.secret_access_key.value());
      }

      return GetClusterCredentials(settings, aws_credentials, credentials, error);

    } catch (const std::exception& e) {
      SetError(error, "%s%s",
               "[aws] Exception during Redshift IAM authentication: ", e.what());
      return Status::IO("Exception during Redshift IAM authentication: ", e.what());
    }
  }

 private:
  Status GetCredentialsFromProfile(const std::string& profile_name,
                                   Aws::Auth::AWSCredentials& credentials,
                                   struct AdbcError* error) const {
    // Create credentials provider for the specified profile
    auto credentials_provider =
        Aws::MakeShared<Aws::Auth::ProfileConfigFileAWSCredentialsProvider>(
            "RedshiftIAMAuth", profile_name.c_str());

    // Get credentials from the provider
    credentials = credentials_provider->GetAWSCredentials();
    if (credentials.GetAWSAccessKeyId().empty() ||
        credentials.GetAWSSecretKey().empty()) {
      SetError(error, "%s%s", "[aws] Failed to get AWS credentials from profile: ",
               profile_name.c_str());
      return Status::IO("Failed to get AWS credentials from profile");
    }

    return Status::Ok();
  }

  Status GetClusterCredentials(const AwsAuthSettings& settings,
                               const Aws::Auth::AWSCredentials& aws_credentials,
                               RedshiftCredentials& credentials,
                               struct AdbcError* error) const {
    Aws::Client::ClientConfiguration config;
    if (settings.region.has_value()) {
      config.region = settings.region.value();
    }

    Aws::Redshift::RedshiftClient redshift_client(aws_credentials, config);
    Aws::Redshift::Model::GetClusterCredentialsRequest request;

    // Strip 'IAM:' prefix if provided
    std::string db_user = settings.user.value_or("awsuser");
    if (db_user.size() >= 4 && db_user.substr(0, 4) == "IAM:") {
      db_user = db_user.substr(4);
    }

    request.SetClusterIdentifier(settings.cluster_id.value_or(""));
    request.SetDbUser(db_user);
    request.SetDbName(settings.database);

    auto outcome = redshift_client.GetClusterCredentials(request);
    if (!outcome.IsSuccess()) {
      SetError(error, "%s%s", "[libpq] Failed to get cluster credentials: ",
               outcome.GetError().GetMessage().c_str());
      return Status::IO("Failed to get cluster credentials: %s",
                        outcome.GetError().GetMessage().c_str());
    }

    // Extract the credentials
    const auto& result = outcome.GetResult();
    credentials.db_user = result.GetDbUser();
    credentials.db_password = result.GetDbPassword();
    if (result.GetExpiration().WasParseSuccessful()) {
      credentials.expiration =
          result.GetExpiration().ToGmtString(Aws::Utils::DateFormat::ISO_8601);
    }

    return Status::Ok();
  }
};

AwsAuthClient::AwsAuthClient() : pimpl_(std::make_unique<Impl>()) {}
AwsAuthClient::~AwsAuthClient() = default;

Status AwsAuthClient::GetRedshiftCredentials(const AwsAuthSettings& settings,
                                             RedshiftCredentials& credentials,
                                             struct AdbcError* error) const {
  return pimpl_->GetRedshiftCredentials(settings, credentials, error);
}
}  // namespace adbcpq

File: go/adbc/driver/redshift/redshift_auth.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package redshift

import (
	"context"
	"fmt"

	"github.com/apache/arrow-adbc/go/adbc"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/ssooidc"
	"github.com/aws/aws-sdk-go-v2/service/sts"
	ststypes "github.com/aws/aws-sdk-go-v2/service/sts/types"
)

const (
	// RFC 8693 OAuth 2.0 Token Exchange grant type, used to exchange the
	// external IdP JWT for an IAM Identity Center identity context.
	idcTokenExchangeGrantType = "urn:ietf:params:oauth:grant-type:token-exchange"
	// Subject token type for an OIDC JWT (RFC 8693).
	idcSubjectTokenTypeJWT = "urn:ietf:params:oauth:token-type:jwt"
	// The AWS-managed context provider ARN for IAM Identity Center trusted
	// identity propagation, supplied to sts:AssumeRole via ProvidedContexts.
	idcContextProviderArn = "arn:aws:iam::aws:contextProvider/IdentityCenter"
	// Session name used for the identity-enhanced assumed-role session.
	idcRoleSessionName = "adbc-redshift-identity-center"
)

// validate ensures the required Identity Center parameters are present.
func (a *IdentityCenterAuth) validate() error {
	missing := func(field string) error {
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg:  fmt.Sprintf("IAM Identity Center auth requires `%s` to be set", field),
		}
	}
	if a == nil || a.token == "" {
		return missing(OptionStringIdCToken)
	}
	// token_type defaults to EXT_JWT; if set it is validated at option time too.
	if a.tokenType != "" && a.tokenType != OptionValueIdCTokenTypeExtJWT {
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg: fmt.Sprintf(
				"unsupported IdC token type `%s`; only `%s` is supported",
				a.tokenType, OptionValueIdCTokenTypeExtJWT),
		}
	}
	if a.clientID == "" {
		return missing(OptionStringIdCClientId)
	}
	if a.roleArn == "" {
		return missing(OptionStringIdCRoleArn)
	}
	return nil
}

// resolveRegion returns the region for the Identity Center endpoints, falling
// back to the driver's AWS region.
func (a *IdentityCenterAuth) resolveRegion(fallback string) string {
	if a.region != "" {
		return a.region
	}
	return fallback
}

// idcAssumeRoleProvider is an aws.CredentialsProvider that produces
// identity-enhanced credentials by assuming a role with an IAM Identity Center
// identity context. Wrapping it in aws.NewCredentialsCache lets long-lived
// connections transparently refresh the assumed-role session on expiry.
type idcAssumeRoleProvider struct {
	stsClient       *sts.Client
	roleArn         string
	identityContext string
}

func (p *idcAssumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
	out, err := p.stsClient.AssumeRole(ctx, &sts.AssumeRoleInput{
		RoleArn:         aws.String(p.roleArn),
		RoleSessionName: aws.String(idcRoleSessionName),
		ProvidedContexts: []ststypes.ProvidedContext{{
			ProviderArn:      aws.String(idcContextProviderArn),
			ContextAssertion: aws.String(p.identityContext),
		}},
	})
	if err != nil {
		return aws.Credentials{}, adbc.Error{
			Code: adbc.StatusUnauthenticated,
			Msg:  fmt.Sprintf("IAM Identity Center AssumeRole failed: %s", err.Error()),
		}
	}
	c := out.Credentials
	if c == nil {
		return aws.Credentials{}, adbc.Error{
			Code: adbc.StatusUnauthenticated,
			Msg:  "IAM Identity Center AssumeRole returned no credentials",
		}
	}
	return aws.Credentials{
		AccessKeyID:     aws.ToString(c.AccessKeyId),
		SecretAccessKey: aws.ToString(c.SecretAccessKey),
		SessionToken:    aws.ToString(c.SessionToken),
		Source:          "RedshiftIdentityCenter",
		CanExpire:       c.Expiration != nil,
		Expires:         aws.ToTime(c.Expiration),
	}, nil
}

// identityCenterCredentialsProvider implements the IAM Identity Center trusted
// identity propagation flow for programmatic access:
//
//  1. Exchange the external IdP JWT (EXT_JWT) for an Identity Center identity
//     context via ssooidc:CreateTokenWithIAM (RFC 8693 token exchange).
//  2. Assume the configured IAM role, passing the identity context through
//     sts:AssumeRole ProvidedContexts, to obtain identity-enhanced AWS
//     credentials.
//
// The resulting credentials carry the user's Identity Center identity and can
// be used with the Redshift Data API when the cluster is configured for IAM
// Identity Center. This mirrors dbt-redshift's `oauth_token_identity_center`
// (redshift_connector `IdpTokenAuthPlugin`, token_type=EXT_JWT) flow.
func (c *connectionImpl) identityCenterCredentialsProvider(ctx context.Context) (aws.CredentialsProvider, error) {
	if err := c.idcAuth.validate(); err != nil {
		return nil, err
	}
	region := c.idcAuth.resolveRegion(c.awsRegion)

	// Base config only reaches the public ssooidc/sts endpoints; the token
	// exchange itself is authenticated by the external JWT, not AWS creds.
	baseCfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
	if err != nil {
		return nil, err
	}

	oidcClient := ssooidc.NewFromConfig(baseCfg)
	tokenOut, err := oidcClient.CreateTokenWithIAM(ctx, &ssooidc.CreateTokenWithIAMInput{
		ClientId:         aws.String(c.idcAuth.clientID),
		GrantType:        aws.String(idcTokenExchangeGrantType),
		SubjectToken:     aws.String(c.idcAuth.token),
		SubjectTokenType: aws.String(idcSubjectTokenTypeJWT),
	})
	if err != nil {
		return nil, adbc.Error{
			Code: adbc.StatusUnauthenticated,
			Msg:  fmt.Sprintf("IAM Identity Center token exchange failed: %s", err.Error()),
		}
	}
	if tokenOut.AwsAdditionalDetails == nil || aws.ToString(tokenOut.AwsAdditionalDetails.IdentityContext) == "" {
		return nil, adbc.Error{
			Code: adbc.StatusUnauthenticated,
			Msg:  "IAM Identity Center token exchange returned no identity context",
		}
	}

	provider := &idcAssumeRoleProvider{
		stsClient:       sts.NewFromConfig(baseCfg),
		roleArn:         c.idcAuth.roleArn,
		identityContext: aws.ToString(tokenOut.AwsAdditionalDetails.IdentityContext),
	}
	return aws.NewCredentialsCache(provider), nil
}
